Add extra support for Projects

ker-an opened the pull request:
# NOTE FROM ADAM: 

### Friday, 5th April 2019: I have merged this branch into mine: https://github.com/Financial-Times/github/tree/refactor-projects

### So no need to merge this PR. Thanks for the code! 

_________________
Previously, the GitHub tooling helper could only create a Project for an organisation. This PR allows one  to create a Project for either an organisation, repository or GitHub user.

Closes #26

sjparkinson commented on src/commands/create-project.js:
```diff
@@ -16,33 +16,46 @@ const builder = yargs => {
+			repo: ["org", "user"],
+			user: ["org", "repo"]
+		})
+		.middleware((argv, yargs) => {
```
From a yargs documentation dive 🏊‍♂️, I think it'd be better to use the yargs `check` function rather than middleware.

http://yargs.js.org/docs/#api-checkfn-globaltrue

> Check that certain conditions are met in the provided arguments.
> 
> `fn` is called with two arguments, the parsed `argv` hash and an array of options and their aliases.
> 
> If `fn` throws or returns a non-truthy value, show the thrown error, usage information, and exit.

So this handles a lot of what is being manually defined in the middleware.

sjparkinson replied:
Might be useful to reference, https://stackoverflow.com/a/54868875.

> yargs demand one of builder option

ker-an replied:
👍 Done! Thanks for the feedback, I've pushed the changes and I think I've addressed all of your points 😄

sjparkinson commented on src/lib/github/create-project.js:
```diff
@@ -1,12 +1,52 @@
-
-		return result.data;
+	return async ({ org, repo, user, target, name, description }) => {
+		if (org) {
```
Worth adding a comment in here highlighting that it's guaranteed one of `org`, `repo`, or `user` will be defined.

sjparkinson commented on src/lib/github/create-project.js:
```diff
@@ -1,12 +1,52 @@
+			 *
+			 * @see https://octokit.github.io/rest.js/#api-Projects-createForRepo
+			 */
+			const string = target.split('/');
```
```suggestion
			const [owner, repo] = target.split('/');
```

sjparkinson replied:
As a variable name, `string` is probably best avoided.

ker-an replied:
Yah I wasn't happy with it, but I'm having such a hard time coming up with a better variable name.

ker-an commented:
@sjparkinson If there's an error creating a project, right now we throw a new error:

```
Error: Creating a project failed. Response: HttpError: Not Found.
    at createProjectError (/Users/keran.braich/Sites/github/src/commands/create-project.js:60:9)
    at process._tickCallback (internal/process/next_tick.js:68:7)
```

If we didn't throw a new error, GitHub's response is:

```
{ HttpError: Not Found
    at response.text.then.message (/Users/keran.braich/Sites/github/node_modules/@octokit/request/lib/request.js:56:27)
    at process._tickCallback (internal/process/next_tick.js:68:7)
  name: 'HttpError',
  status: 404,
  headers:
   { 'access-control-allow-origin': '*',
     'access-control-expose-headers':
      'ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type',
     connection: 'close',
     'content-encoding': 'gzip',
     'content-security-policy': 'default-src \'none\'',
     'content-type': 'application/json; charset=utf-8',
     date: 'Tue, 26 Mar 2019 13:41:05 GMT',
     'referrer-policy': 'origin-when-cross-origin, strict-origin-when-cross-origin',
     server: 'GitHub.com',
     status: '404 Not Found',
     'strict-transport-security': 'max-age=31536000; includeSubdomains; preload',
     'transfer-encoding': 'chunked',
     'x-accepted-oauth-scopes': 'repo',
     'x-content-type-options': 'nosniff',
     'x-frame-options': 'deny',
     'x-github-media-type':
      'github.inertia-preview; format=json, github.inertia-preview; format=json',
     'x-github-request-id': 'E2FE:77B0:7900A5:F1EB5B:5C9A2BF0',
     'x-oauth-scopes': 'repo',
     'x-ratelimit-limit': '5000',
     'x-ratelimit-remaining': '4992',
     'x-ratelimit-reset': '1553611265',
     'x-xss-protection': '1; mode=block' },
  request:
   { method: 'POST',
     url: 'https://api.github.com/orgs/financial-butts/projects',
     headers:
      { accept:
         'application/vnd.github.inertia-preview+json,application/vnd.github.inertia-preview+json',
        'user-agent': 'octokit.js/16.20.0 Node.js/10.15.0 (macOS Mojave; x64)',
        authorization: 'token [REDACTED]',
        'content-type': 'application/json; charset=utf-8' },
     body: '{"name":"project name"}',
     request: { validate: [Object] } },
  documentation_url:
   'https://developer.github.com/v3/projects/#create-an-organization-project' }
```

simonplend commented on src/commands/close-project.js:
```diff
@@ -0,0 +1,50 @@
+	});
+
+	console.log(openProjects);
+};
```
To be consistent with the existing commands, this command should implement...

- Support for the `--json` option
- Consistent output by using the `print-output` helper

See [src/commands/create-project.js](https://github.com/Financial-Times/github/blob/master/src/commands/create-project.js) for an example.

ker-an replied:
Sounds good! I will do that 👍

simonplend commented on src/commands/close-project.js:
```diff
@@ -0,0 +1,50 @@
+};
+
+/**
+ * Close a project.
```
I'm confused... the commit message says "List all open projects for an org" :thinking:

simonplend commented on src/commands/close-project.js:
```diff
@@ -0,0 +1,50 @@
+};
+
+module.exports = {
+	command: "projects:close <path>",
```
I'm confused... the handler is listing projects, not closing a project :thinking:

ker-an replied:
Yes, sorry I should have explained. In order to close a project, we need a project ID which as we learned isn't very easy. It's not as simple as getting it from the URL. The way that Sam and I discussed it was that we would:

1. Extract the project type (e.g. repos, orgs, users) and name (e.g. financial-times-sandbox) from the URL passed in through the command line.

2. Use the GitHub API to list all of the `open` projects under that account

3. Find the project to close and project ID by using the URL originally passed in as it will be somewhere in the GitHub API response.

4. Use the GitHub API and project ID to close the project.

Yah, it's not pretty but I couldn't think of anything else. Any suggestions for improvement on this process so welcome!

ker-an replied:
So far I have done points 1 and 2.

simonplend commented on src/lib/github.js:
```diff
@@ -1,9 +1,9 @@
-const Octokit = require("@octokit/rest");
```
Woah! What happened to this file?

ker-an replied:
Oh dear, I was using single quotes and then was like oh double quotes, why don't I change you to single quotes for consistency.

simonplend replied:
Ahh, we should figure out how to keep quotes consistent. See: https://github.com/Financial-Times/github/commit/b5325968bd7577aa47932e9e206e3f028f634cc7

ker-an replied:
haha I must have missed the memo about double quotes over single quotes, good to know!

simonplend replied:
Well, that's a good point. This project (and a few of our other new ones) have been formatted with Prettier, which by default uses whichever quotes it things are best, depending on the contents of the string. I'm personally fine with this, but it's definitely different to [how things are in Next-land](https://github.com/Financial-Times/n-gage/blob/master/dotfiles/.eslintrc.js#L28) (quotes must always be single).

simonplend commented on src/commands/close-project.js:
```diff
@@ -0,0 +1,50 @@
+
+	return baseOptions(yargs)
+		.positional('path', {
+            describe: 'URL path to project',
```
```suggestion
            describe: 'Project URL',
```

ker-an commented:
Closing this PR and deleting this branch as the code as been merged into @adambraimbridge `refactor-projects` branch. I am going to work off of the latest version that incorporates Adam's changes. I've addressed all the PR comments above so I think it's okay to close this 😄
